pkg/data/account_test.go

```diff
@@ -0,0 +1,251 @@
+			dynamoOutput: &dynamodb.DeleteItemOutput{
+				Attributes: map[string]*dynamodb.AttributeValue{},
+			},
+			expectedErr: errors.NewInternalServer("delete failed for account \"abc123\"", gErrors.New("failure")),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mockDynamo := awsmocks.DynamoDBAPI{}
+
+			mockDynamo.On("DeleteItem", mock.MatchedBy(func(input *dynamodb.DeleteItemInput) bool {
+				return (*input.TableName == "Accounts" &&
+					*input.Key["Id"].S == *tt.account.ID)
+			})).Return(
+				tt.dynamoOutput, tt.dynamoErr,
+			)
+			accountData := &Account{
+				DynamoDB:  &mockDynamo,
+				TableName: "Accounts",
+			}
+
+			err := accountData.DeleteAccount(&tt.account)
+			assert.True(t, errors.Is(err, tt.expectedErr))
+		})
+	}
+
+}
+
+func TestUpdate(t *testing.T) {
+	tests := []struct {
+		name              string
+		account           model.Account
+		dynamoErr         error
+		expectedErr       error
+		oldLastModifiedOn *int64
+	}{
+		{
+			name: "update",
+			account: model.Account{
+				ID:             ptrString("abc123"),
+				Status:         model.AccountStatusReady.AccountStatusPtr(),
+				LastModifiedOn: ptrInt64(1573592058),
+				AdminRoleArn:   ptrString("test:Arn"),
+			},
+			oldLastModifiedOn: ptrInt64(1573592057),
+			dynamoErr:         nil,
+			expectedErr:       nil,
+		},
+		{
+			name: "create",
+			account: model.Account{
+				ID:             ptrString("abc123"),
+				Status:         model.AccountStatusReady.AccountStatusPtr(),
+				LastModifiedOn: ptrInt64(1573592058),
+				AdminRoleArn:   ptrString("test:Arn"),
+			},
+			dynamoErr:   nil,
+			expectedErr: nil,
+		},
+		{
+			name: "conditional failure",
+			account: model.Account{
+				ID:             ptrString("abc123"),
+				Status:         model.AccountStatusReady.AccountStatusPtr(),
+				LastModifiedOn: ptrInt64(1573592058),
+				AdminRoleArn:   ptrString("test:Arn"),
+			},
+			oldLastModifiedOn: ptrInt64(1573592057),
+			dynamoErr:         awserr.New("ConditionalCheckFailedException", "Message", fmt.Errorf("Bad")),
+			expectedErr:       errors.NewConflict("account", "abc123", fmt.Errorf("unable to update account with LastModifiedOn=\"1573592058\"")),
+		},
+		{
+			name: "other dynamo error",
+			account: model.Account{
+				ID:             ptrString("abc123"),
+				Status:         model.AccountStatusReady.AccountStatusPtr(),
+				LastModifiedOn: ptrInt64(1573592058),
+				AdminRoleArn:   ptrString("test:Arn"),
+			},
+			oldLastModifiedOn: ptrInt64(1573592057),
+			dynamoErr:         gErrors.New("failure"),
+			expectedErr:       errors.NewInternalServer("update failed for account \"abc123\"", gErrors.New("failure")),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mockDynamo := awsmocks.DynamoDBAPI{}
+
+			mockDynamo.On("PutItem", mock.MatchedBy(func(input *dynamodb.PutItemInput) bool {
+				if tt.oldLastModifiedOn == nil {
+					return (*input.TableName == "Accounts" &&
```

Comment:
You can actually throw `assert.Equal()` statements into these mock callbacks, instead of having to write out big conditionals like this. It gives you a nicer messages, then, if any of your conditions are false.

Just don't use `require.Equal()` -- it breaks your test in a gross way.